File: rust/bindings/uniffi/python_proof/proof.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "generated", "python"))
from cypher_ffi import Sender, Receiver, generate_phrase

phrase = generate_phrase()
original = b"hello from the cypher python binding"
frames = Sender(phrase, False).frames(original, "greeting.txt")
r = Receiver(phrase, False)
for f in frames:
    r.push_frame(f)
assert r.is_complete(), "receiver did not complete"
out = r.data()
ok = out == original and r.name() == "greeting.txt"
print(f'frames={len(frames)} decoded="{out.decode()}" name="{r.name()}"')
print("PASS: Python binding round-trip byte-exact (thin UniFFI over the Rust core)" if ok else "FAIL")

File: rust/bindings/wasm/proof.mjs
import assert from "node:assert";
import { Sender, Receiver, generate_phrase } from "./pkg/cypher_wasm.js";

const phrase = generate_phrase();
const original = "hello from the cypher wasm sdk";
const filename = "greeting.txt";

// --- Happy path: real random keys, real clock, byte-exact round-trip ---
const s = new Sender(phrase, false);
const frames = s.frames(new TextEncoder().encode(original), filename);
assert(frames.length > 0, "sender produced no frames");

const r = new Receiver(phrase, false);
for (const f of frames) r.push_frame(f);
assert(r.is_complete(), "receiver did not complete");

const out = r.data();
const decoded = new TextDecoder().decode(out);
assert.strictEqual(decoded, original, "payload mismatch");
assert.strictEqual(r.name(), filename, "name mismatch");

// --- Negative: wrong phrase must not decode ---
const wrong = new Receiver(generate_phrase(), false);
for (const f of frames) {
  try {
    wrong.push_frame(f);
  } catch (_) {
    /* auth/decode errors are expected and fine */
  }
}
assert(!wrong.is_complete(), "wrong-phrase receiver completed - PSK not enforced");

console.log(`PASS: decoded=${JSON.stringify(decoded)} name=${JSON.stringify(r.name())} frames=${frames.length}`);
console.log("PASS: wrong-phrase receiver did NOT complete");
